src/chains/ethereum/ethereum/src/helpers/trie.ts

```diff
@@ -1,33 +1,50 @@
-import { Quantity } from "@ganache/utils";
-import { LevelUp } from "levelup";
-import { SecureTrie } from "merkle-patricia-tree";
+import { keccak, Quantity } from "@ganache/utils";
+import { Trie } from "@ethereumjs/trie";
 import Blockchain from "../blockchain";
+import { TrieDB } from "../trie-db";
 
-export class GanacheTrie extends SecureTrie {
+const keyHashingFunction = (msg: Uint8Array) => {
+  return keccak(Buffer.from(msg.buffer, msg.byteOffset, msg.length));
+};
+
+export class GanacheTrie extends Trie {
   public readonly blockchain: Blockchain;
+  /**
+   * The database that's returned from this.database() does not have all of
+   * the types of the original input database because ethereumjs doesn't use
+   * generics for their types in the underlying `CheckpointDB`. So, we store the
+   * original db on our trie so we can access those types.
+   */
+  public readonly db: TrieDB;
```

Comment:
I'm a little uncomfortable with maintaining a second reference here. We _could_ implement an accessor that returns the underlying property, but it appears to be slower. Although I don't know how much performance would matter here.

```typescript
public get db(): TrieDB {
  return super._db as TrieDB;
}
```

:shrug:

Reply:
I'm not super opinionated on this. @davidmurdoch, @tenthirtyone, thoughts?

Reply:
You could say the same about the inverse though. The `._db` property is protected, so it's _intended_ to be used by sub classes.

I can imagine a scenario where `super._db` is modified in the constructor to be different to the value passed as an argument, making it really difficult to debug issues, where we expect the super class to be acting on the database instance referred to by our own `.db` property.

But, again 🤷 , I'm not super opinionated on it, so happy if you guys disagree.

Reply:
@jeffsmale90 We tried to get ethereumjs to make the db generic so it returns the original type it was given, but they do not won't want to support that. So I think the argument your presented _against_ using `._db` is very valid.